airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py

```diff
@@ -0,0 +1,278 @@
+        # of which were already synced using the Concurrent CDK
+        filtered_catalog = self._remove_concurrent_streams_from_catalog(catalog=catalog, concurrent_stream_names=concurrent_stream_names)
+        yield from super().read(logger, config, filtered_catalog, state)
+
+    def check(self, logger: logging.Logger, config: Mapping[str, Any]) -> AirbyteConnectionStatus:
+        # I don't think should need to be overwritten because I added all_streams() and the underlying check ->
+        # CheckStream.check_connection() should invoke all_streams(). It should also sort of be a no-op because
+        # we've effectively deprecated availability_strategy as a concept which is all the ConnectionChecker does
+        return super().check(logger=logger, config=config)
+
+    def discover(self, logger: logging.Logger, config: Mapping[str, Any]) -> AirbyteCatalog:
+        return AirbyteCatalog(streams=[stream.as_airbyte_stream() for stream in self.all_streams(config=config)])
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        """
+        Returns the list of streams that can be run synchronously in the Python CDK.
+
+        NOTE: For ConcurrentDeclarativeSource, this method only returns synchronous streams because it usage is invoked within the
+        existing Python CDK. Streams that support concurrency are started from read().
+        """
+        return self._synchronous_streams
+
+    def all_streams(self, config: Mapping[str, Any]) -> List[Union[AbstractStream, Stream]]:
+        return self._synchronous_streams + self._concurrent_streams
+
+    def _separate_streams(self, config: Mapping[str, Any]) -> (List[AbstractStream], List[Stream]):
+        concurrent_streams: List[AbstractStream] = []
+        synchronous_streams: List[Stream] = []
+
+        state_manager = ConnectorStateManager(state=self._state)
+
+        for declarative_stream in super().streams(config=config):
+            # Some low-code sources use a combination of DeclarativeStream and regular Python streams. We can't inspect
+            # these legacy Python streams the way we do low-code streams to determine if they are concurrent compatible,
+            # so we need to treat them as synchronous
+            if isinstance(declarative_stream, DeclarativeStream):
+                declarative_cursor_attributes = self._get_cursor_attributes(declarative_stream=declarative_stream, config=config)
+                if declarative_cursor_attributes:
+                    stream_state = state_manager.get_stream_state(
+                        stream_name=declarative_stream.name, namespace=declarative_stream.namespace
+                    )
+
+                    connector_state_converter = CustomOutputFormatConcurrentStreamStateConverter(
+                        datetime_format=declarative_cursor_attributes.datetime_format,
+                        is_sequential_state=False,
```

Comment:
By disabling `is_sequential_state`, we will automatically move connectors from their sequential state to per-partition concurrent state. This is the ideal end goal we want, however, if we were to revert the connector to a previous version that doesn't use the concurrent CDK, the state message would not be compatible.

It feels like to de-risk this a little bit on the early release, we should set this to true so we accept sequential state (i.e. `{"updated_at": "2024-12-12"}` and also emit it back to the platform as such

Reply:
do you think we could add a mechanism to handle this situation by using the lowest value to set the sequential state if we revert the changes?